signer/src/bitcoin/validation.rs

```diff
@@ -91,10 +104,121 @@ pub fn is_unique(package: &[TxRequestIds]) -> bool {
+        &self,
+        ctx: &C,
+        btc_ctx: &BitcoinTxContext,
+    ) -> Result<ValidationCache, Error>
+    where
+        C: Context + Send + Sync,
+    {
         if !is_unique(&self.request_package) {
             return Err(Error::DuplicateRequests);
         }
+        let cache = self.fetch_all_reports(ctx, btc_ctx).await?;
+
+        self.validate_max_mintable(ctx, &cache).await?;
+
+        Ok(cache)
+    }
+
+    async fn fetch_all_reports<C>(
+        &self,
+        ctx: &C,
+        btc_ctx: &BitcoinTxContext,
+    ) -> Result<ValidationCache, Error>
+    where
+        C: Context + Send + Sync,
+    {
+        let db = ctx.get_storage();
+        let mut cache = ValidationCache {
+            sbtc_limits: ctx.get_emily_client().get_limits().await?,
+            ..Default::default()
+        };
+
+        // Fetch all deposit reports and votes
+        for requests in &self.request_package {
+            for outpoint in &requests.deposits {
+                let txid = outpoint.txid.into();
+                let output_index = outpoint.vout;
+
+                let report_future = db.get_deposit_request_report(
+                    &btc_ctx.chain_tip,
+                    &txid,
+                    output_index,
+                    &btc_ctx.signer_public_key,
+                );
+                let Some(report) = report_future.await? else {
+                    return Err(InputValidationResult::Unknown.into_error(btc_ctx));
+                };
+
+                let votes = db
+                    .get_deposit_request_signer_votes(&txid, output_index, &btc_ctx.aggregate_key)
+                    .await?;
+
+                cache
+                    .deposit_reports
+                    .insert((&outpoint.txid, output_index), (report, votes));
+            }
+
+            // Fetch all withdrawal reports and votes
+            for id in &requests.withdrawals {
+                let report = db.get_withdrawal_request_report(
+                    &btc_ctx.chain_tip,
+                    id,
+                    &btc_ctx.signer_public_key,
+                );
+                let Some(report) = report.await? else {
+                    return Err(WithdrawalValidationResult::Unknown.into_error(btc_ctx));
+                };
+
+                let votes = db
+                    .get_withdrawal_request_signer_votes(id, &btc_ctx.aggregate_key)
+                    .await?;
+
+                cache.withdrawal_reports.insert(id, (report, votes));
+            }
+        }
+        Ok(cache)
+    }
+
+    async fn validate_max_mintable<'a, C>(
+        &self,
+        ctx: &C,
+        cache: &ValidationCache<'a>,
+    ) -> Result<(), Error>
+    where
+        C: Context + Send + Sync,
+    {
+        let sbtc_supply = ctx
+            .get_stacks_client()
+            .get_sbtc_total_supply(&ctx.config().signer.deployer)
+            .await?;
+        let total_cap = cache.sbtc_limits.total_cap().unwrap_or(Amount::MAX_MONEY);
```

Comment:
I'd add some comments here explaining that if the limit is `None` in Emily then we use `MAX_MONEY` to simulate no cap, etc. (it might not be obvious to a future reader why we're doing this)

Reply:
comments added wherever we call an `unwrap_or` in in https://github.com/stacks-network/sbtc/pull/967/commits/fe55499a492a0b1b5f94ac40f2c1438f50478ef3